### DISCUSSION TITLE
Need help understanding invalidation caches

### DISCUSSION THREAD

**Discussion by vmgbrown** [infinispan/infinispan#10821] [Q&A] [answered]
Hi all,

I'm trying to resolve a performance issue with a Java application and would like to get a better understanding of how invalidation caches work. The application is deployed in WildFly 26, which appears to include Infinispan 12.1.7. We are using Infinispan invalidation caches as a 2nd-level cache for Hibernate in this application.

The behavior we are seeing is that one of the nodes in a multi-node system occasionally seems to lose all cache data for a particular entity type. When this happens, the data is requested again from the database, which has a noticeable performance impact. 

From what I have read, invalidation caches do not share data but simply represent the state of an entity by key. What I don't quite understand is when/how the cache entries are invalidated. 

For example, let's say there are 2 nodes A and B and the system has just started (i.e. the caches on both nodes are empty). Node A looks up entity 1, which is retrieved from the database and added to the node A cache. Subsequent lookups of entity 1 on node A will presumably use this value.

Next, node B looks up entity 1. Because node B has not yet requested entity 1, I assume it must again be retrieved from the database and added to the node B cache. If so, is an invalidation message sent at this point, notifying A that it should remove entity 1 from its cache? I wouldn't think so, but I haven't been able to find anything definitive to prove it.

Clearly if either node A or B make and save changes to entity 1, then an invalidation message must be sent. But it seems like invalidating entries as they are added to the cache would completely defeat the purpose of using an invalidation cache at all.

The cache in question has a `heap-memory` value of 100. I am wondering if this could be the source of the problem. Assuming the cache uses an LRU algorithm, it seems plausible that 100 new entries could be added to the cache during execution of the problematic logic, displacing the initially cached entries and requiring additional database lookups. Would appreciate it if someone could confirm my understanding of this behavior.

Thanks,
Greg

**Comment by vmgbrown** [comment]
Related question - for the `cache-container` configuration below, does the `heap-memory` size value apply to all entries in the cache, or only to entries of a specific type?

```
<cache-container name="..." marshaller="JBOSS" modules="org.infinispan.hibernate-cache">
  <invalidation-cache name="entity" remote-timeout="120000">
      <heap-memory size="100"/>
      <expiration max-idle="300000" lifespan="300000"/>
  </invalidation-cache>
</cache-container>
```
In other words, if the cache container is used by _n_ entity types, is the total cache size _n_ * _size_, or is it just _size_? I assume _size_, but would like to confirm.

**Reply by wburns** [reply]
First, the size is specific only to that cache. Then inside the cache the size does not look at any types in it, it is evicted solely based on the eviction algorithms choice.

**Reply by vmgbrown** [reply]
Thank you. I am still struggling a bit to understand how invalidation caches work. It seems like they essentially behave like local caches, only they evict entries if an invalidation message is received from another node. 

In other words, two nodes using an invalidation cache might both retrieve the same entity from the database independently. This is in contrast to a replicated cache, where the second node would automatically receive a copy of the entity as soon as the first node had read it. "Invalidation" only seems to apply if one node or the other modifies an entity that both nodes have cached. Is that correct?

**Reply by tristantarrant** [reply]
Correct. Quoting https://infinispan.org/docs/stable/titles/configuring/configuring.html#invalidation-caches_clustered-caches

> Sometimes the application reads a value from the external store and wants to write it to the local cache, without removing it from the other nodes. To do this, it must call Cache.putForExternalRead(key, value) instead of Cache.put(key, value).

**Reply by vmgbrown** [reply]
I did see that in the documentation. However, my application does not access the cache directly - we are using it as an L2 cache for Hibernate. So essentially I am trying to understand how Hibernate uses Infinispan. 

Thank you for the information.

**Reply by tristantarrant** [reply]
The 2LC code uses "put for external read" under the hood when data is loaded from the DB. See https://github.com/infinispan/infinispan/blob/main/hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/access/InvalidationCacheAccessDelegate.java#L86

**Reply by vmgbrown** [reply]
@tristantarrant That's very helpful, thanks.

Do you have any thoughts on the Cache vs. CacheManager question below?

**Comment by vmgbrown** [comment]
In an effort to troubleshoot the issue, I enabled statistics on my cache containers; e.g.
```
<cache-container name="..." statistics-enabled="true">
```
What is the expected result of this setting? I assumed I would be able to view the captured statistics in JMX, but when I started my server it appeared that statistics were still not enabled.

**Reply by wburns** [reply]
Maybe you did not enable exporting the statistics to JMX? https://docs.jboss.org/infinispan/12.0/configdocs/infinispan-config-12.0.html Under JMX the beans are not exported by default.

**Reply by vmgbrown** [reply]
Additional information. In JMX (Hawtio), I see the following:
* org.wildfly.clustering.infinispan
  * Cache
  * CacheManager

Under "CacheManager", I see my named cache container, and I see that statistics are enabled (although all the values are 0). 

Under "Cache", I see what appears to be a list of entity types. Expanding an entity type shows a list of cache containers that are presumably associated with the entity. Expanding the same container node here, I see that statistics are NOT enabled.

Is this expected? What is the difference between Cache and CacheManager here?

**Reply by vmgbrown** [reply]
More specifically, this is a simplified version of what I am seeing in JMX:
* org.wildfly.clustering.infinispan
  * Cache
    * "com.foo.MyEntityType(local)"
      * "MyCacheContainer"
        * Statistics
          * Statistics enabled - **false**
  * CacheManager
    * "MyCacheContainer"
      * CacheContainerStats
        * Statistics enabled - **true**

I assume that the values under CacheManager would reflect the statistics for the entire named container. So given a `heap-size` of 100, the "number of entries" in the container should never be more than 100, correct?

I would guess that the values under Cache show the statistics for a particular entity within the container. So if I had two entity types associated with my container, I might see behavior like the following:
* Load 50 instances of entity A - cache contains 50 entries of A
* Load 105 instances of entity B - cache contains 100 entries of B

Does that sound right? If so, the stats of Cache/com.foo.A should show 0 entries, and  Cache/com.foo.B should show 100. Does that sound right?

I'd also like to understand why statistics are not enabled for the entity node but are enabled for the manager node. Is this expected? Am I misunderstanding something?

**Reply by wburns** [reply]
Cache manager can hold N caches. For hibernate 2LC it creates a cache for each of the entities that have 2LC enabled.

In this case you have only enabled statistics at the cache manager side but not the actual cache. You would need to enable it on each cache that you want the stats made available.

**Reply by vmgbrown** [reply]
Thanks again for the prompt response. Where can I enable statistics at the individual cache level? I didn't see an obvious way to do this in `standalone.xml`.

Also, to confirm something I had asked about earlier - if a cache container contains _n_ caches (one per entity type), does the container's `heap-memory` size value apply to the entire container, or to each cache within the container? Based on your earlier response, I had assumed it applied to the container:

> the size is specific only to that cache...inside the cache the size does not look at any types in it, it is evicted solely based on the eviction algorithms choice

However, I now see that you said "cache", not "container", so I wanted to make sure I understood.

**Reply by wburns** [reply]
> Thanks again for the prompt response. Where can I enable statistics at the individual cache level? I didn't see an obvious way to do this in standalone.xml.

If you look at the schema I linked earlier each cache has an attribute named `statistics` that is false by default.

> Also, to confirm something I had asked about earlier - if a cache container contains n caches (one per entity type), does the container's heap-memory size value apply to the entire container, or to each cache within the container? Based on your earlier response, I had assumed it applied to the container:

Unfortunately, eviction is done solely on a per cache basis. There is an improvement somewhere to have a cache manager wide eviction, but that is not implemented yet.

**Reply by vmgbrown** [reply]
> If you look at the schema I linked earlier each cache has an attribute named statistics that is false by default.

I tried enabling statistics at the cache level as described in that schema, but it didn't seem to work. I will try again.

> Unfortunately, eviction is done solely on a per cache basis. There is an improvement somewhere to have a cache manager wide eviction, but that is not implemented yet.

All I would like to understand is if a cache container's `heap-memory` size value applies to the container as a whole, or if it applies to each cache in the container. 

Let's say I have two caches associated with this container, one for entity A and another for entity B:

```
<cache-container name="abc" marshaller="JBOSS" modules="org.infinispan.hibernate-cache">
    <local-cache name="entity">
        <heap-memory size="100"/>
    </local-cache>
</cache-container>
```

Will both caches be allowed to store up to 100 entries each, or is the _total_ number of entries in the A and B caches limited to 100?

For example, if my code does this:
* Load 50 instances of entity A
* Load 105 instances of entity B

will cache A now have 0 entries and B 100, or will A contain 50 and B 100?

Thanks again for your help.

**Comment by wburns** [comment]
An invalidation cache only sends invalidations when the cache is explicitly updated via Cache.put, Cache.remove, Cache.replace etc. From your description it sounds like eviction is the reason for the loss of the entry.

**Reply by vmgbrown** [reply]
I would have expected an invalidation to result in an eviction. Is this not correct?

**Reply by wburns** [reply]
Eviction is different than an invalidation. Eviction happens solely due to the container hitting its configured limit. An invalidation happens when an entry is updated in the cache and it sends that message to other nodes. There are different stats and listeners for eviction and invalidation.

**Comment by tristantarrant** [comment, marked as answer]
Values are per-cache. We don't support per-container bounds
